rust_src/src/crypto/mod.rs

```diff
@@ -1,78 +1,467 @@
+           default value of buffer-file-coding-system. */
+        return LispObject::from_raw(buffer.buffer_file_coding_system);
+    }
+    let sscsf = LispObject::from_raw(unsafe { globals.f_Vselect_safe_coding_system_function });
+    if fboundp(sscsf).is_not_nil() {
+        /* Confirm that VAL can surely encode the current region. */
+        return call!(
+            sscsf,
+            LispObject::from_natnum(start_byte as EmacsInt),
+            LispObject::from_natnum(end_byte as EmacsInt),
+            coding_system,
+            LispObject::constant_nil()
+        );
+    }
+    LispObject::constant_nil()
+}
+
+fn get_input_from_string(
+    object: LispObject,
+    string: LispStringRef,
+    start: LispObject,
+    end: LispObject,
+) -> LispObject {
+    let size: ptrdiff_t;
+    let start_byte: ptrdiff_t;
+    let end_byte: ptrdiff_t;
+    let mut start_char: ptrdiff_t = 0;
+    let mut end_char: ptrdiff_t = 0;
+
+    size = string.len_bytes();
+    unsafe {
+        validate_subarray(
+            object.to_raw(),
+            start.to_raw(),
+            end.to_raw(),
+            size,
+            &mut start_char,
+            &mut end_char,
+        );
+    }
+    start_byte = if start_char == 0 {
+        0
+    } else {
+        unsafe { string_char_to_byte(object.to_raw(), start_char) }
+    };
+    end_byte = if end_char == size {
+        string.len_bytes()
+    } else {
+        unsafe { string_char_to_byte(object.to_raw(), end_char) }
+    };
+    if start_byte == 0 && end_byte == size {
+        object
+    } else {
+        LispObject::from_raw(unsafe {
+            make_specified_string(
+                string.const_sdata_ptr().offset(start_byte),
+                -1 as ptrdiff_t,
+                end_byte - start_byte,
+                string.is_multibyte(),
+            )
+        })
+    }
+}
+
+fn get_input_from_buffer(
+    buffer: LispBufferRef,
+    start: LispObject,
+    end: LispObject,
+    start_byte: &mut ptrdiff_t,
+    end_byte: &mut ptrdiff_t,
+) -> LispObject {
+    let prev_buffer = unsafe { (*current_thread).m_current_buffer };
+    unsafe { record_unwind_current_buffer() };
+    unsafe { set_buffer_internal(buffer.as_ptr() as *const _ as *const libc::c_void) };
+    *start_byte = if start.is_nil() {
+        buffer.begv
+    } else {
+        match start.as_number_coerce_marker_or_error() {
+            LispNumber::Fixnum(n) => n as ptrdiff_t,
+            LispNumber::Float(n) => n as ptrdiff_t,
+        }
+    };
+    *end_byte = if end.is_nil() {
+        buffer.zv
+    } else {
+        match end.as_number_coerce_marker_or_error() {
+            LispNumber::Fixnum(n) => n as ptrdiff_t,
+            LispNumber::Float(n) => n as ptrdiff_t,
+        }
+    };
+    if start_byte > end_byte {
+        std::mem::swap(start_byte, end_byte);
+    }
+    //if !(buffer.begv <= start_byte && end_byte <= buffer.zv) {
+    //    args_out_of_range(start, end);
+    //}
```

Comment:
Oops, I simply forgot about it. We don't yet have it declared in remacs_sys, and it's not yet included amongst the new macros either.